tsl/src/remote/connection.h

```diff
@@ -116,6 +144,68 @@ extern void remote_connection_stats_reset(void);
 extern RemoteConnectionStats *remote_connection_stats_get(void);
 #endif
 
+/*
+ * Connection functions for COPY mode.
+ */
+extern bool remote_connection_begin_copy(TSConnection *conn, const char *copycmd, bool binary,
+										 TSConnectionError *err);
+extern bool remote_connection_end_copy(TSConnection *conn, TSConnectionError *err);
+extern bool remote_connection_put_copy_data(TSConnection *conn, const char *buffer, size_t len,
+											TSConnectionError *err);
+
+/* Error handling functions for connections */
+extern void remote_connection_get_error(const TSConnection *conn, TSConnectionError *err);
+extern void remote_connection_get_result_error(const PGresult *res, TSConnectionError *err);
+
+/*
+ * The following are macros for emitting errors related to connections or
+ * remote command execution. They need to be macros to preserve the error
+ * context of where they are called (line number, statement, etc.).
+ */
+#define remote_connection_error_elog(err, elevel)                                                  \
+	ereport(elevel,                                                                                \
+			((err)->remote.errcode != 0 ? errcode((err)->remote.errcode) :                         \
+										  errcode((err)->errcode),                                 \
+			 (err)->remote.msg ?                                                                   \
+				 errmsg_internal("[%s]: %s", (err)->nodename, (err)->remote.msg) :                 \
+				 ((err)->connmsg ? errmsg_internal("[%s]: %s", (err)->nodename, (err)->connmsg) :  \
+								   errmsg_internal("[%s]: %s", (err)->nodename, (err)->msg)),      \
+			 (err)->remote.detail ? errdetail_internal("%s", (err)->remote.detail) : 0,            \
+			 (err)->remote.hint ? errhint("%s", (err)->remote.hint) : 0,                           \
+			 (err)->remote.sqlcmd ? errcontext("Remote SQL command: %s", (err)->remote.sqlcmd) :   \
+									0))
```

Comment:
I find `static inline` functions in header files to be perfect for this. You still get the same overhead as a macro since you avoid the function call, but you still get full type checking.

Reply:
In general, I do not worry about function calls. They are quite performant on modern architectures.

Reply:
I still see some value in avoiding them when possible, for various reasons that may be more relevant in IOT devices, e.g.:
- You take advantage of instruction cache locality since you don't have to move the CPU Instruction Pointer.
- You can save some CPU register space and more importantly stack memory space by avoiding the call.
- Virtualised environments (hypervisors) may magnify this kind of overheads.

But you're probably right that the function call overhead itself is kind of irrelevant.

Reply:
This needs to be a macro to capture the context of where the error is through, like line-number, function where error is thrown, etc. If it is not a macro, it believe it will look as if all errors are thrown from the same place.

Reply:
That is right. Ok, then we have to keep the macro.

Reply:
FWIW, there's actually a comment above the macro that explains why it needs to be a macro.

Reply:
One way would be this https://stackoverflow.com/questions/54049362/tracing-static-inline-function but  it is probably not worth the effort